src/auth/downscopedclient.ts

```diff
@@ -0,0 +1,254 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {GaxiosOptions, GaxiosPromise, GaxiosResponse} from 'gaxios';
+
+import {BodyResponseCallback} from '../transporters';
+import {Credentials} from './credentials';
+import {AuthClient} from './authclient';
+
+import {GetAccessTokenResponse, Headers, RefreshOptions} from './oauth2client';
+import * as sts from './stscredentials';
+
+/**
+ * The required token exchange grant_type: rfc8693#section-2.1
+ */
+const STS_GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:token-exchange';
+/**
+ * The requested token exchange requested_token_type: rfc8693#section-2.1
+ */
+const STS_REQUEST_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:access_token';
+/**
+ * The requested token exchange subject_token_type: rfc8693#section-2.1
+ */
+const STS_SUBJECT_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:access_token';
+/** The STS access token exchange end point. */
+const STS_ACCESS_TOKEN_URL = 'https://sts.googleapis.com/v1beta/token';
+
+/**
+ * Offset to take into account network delays and server clock skews.
+ */
+export const EXPIRATION_TIME_OFFSET = 5 * 60 * 1000;
+
+/**
+ * Internal interface for tracking the access token expiration time.
+ */
+interface CredentialsWithResponse extends Credentials {
+  res?: GaxiosResponse | null;
+}
+
+/**
+ * Defines an upper bound of permissions available for a GCP credential.
+ */
+interface CredentialAccessBoundary {
+  accessBoundary: {
+    accessBoundaryRules: AccessBoundaryRule[];
+  };
+}
+
+/** Defines an upper bound of permissions on a particular resource. */
+interface AccessBoundaryRule {
```

Comment:
I forgot to mention. These CAB rule interfaces have to also be exported as they are externally visible. Of course we should do this when we are ready for launch.